### ISSUE TITLE
document how to rebuild caches

### ISSUE THREAD

**Issue by aspiers** [radian-software/straight.el#1024]
I created a new recipe in my local fork of the `melpa` for the `lsp-bridge` package which is not yet in MELPA, and wanted to test installation from that recipe.  However straight did not notice the addition of that recipe, so I couldn't.  For example, `M-x straight-rebuild-recipe` doesn't accept `lsp-bridge` as a valid package name.

What's the correct procedure for achieving this?  I'm guessing this requires some cache rebuilding; if so, it would be great if this was implemented in a convenient user-facing interactive function, and documented accordingly.  Thanks!

**Comment by raxod502** [comment]
Yep, it's definitely a valid request. The existing documentation has most of what you need, but it's not organized in a way that makes it easy to get started. (Sorry!)

> headings like "Where do recipes come from?" and "What does it mean to register a package?" which sound like they are written for people who mainly want to understand the internals of straight.el

Actually, that is intended to be the user facing part. You don't even want to know what the [developer facing part](https://github.com/radian-software/straight.el/issues/51) would look like ;)

Obviously, this can and should be improved, because having a full conceptual understanding of all of the pieces of your package manager should not be a requirement to use it effectively. This is something that I will address when I have time to focus primarily on `straight.el`; in the meantime, any improvements are welcome.

**Comment by aspiers** [comment]
Thanks for the reply!  I would suggest considering _three_ target audiences, rather than two:

1. emacs users who just want to install and use packages
2. developers who want to write and/or improve recipes
3. developers who want to make `straight.el` even better

This issue focuses purely on 2., which I assume would be pretty easy to address since recipes are generally so simple and `straight.el` already has all the mechanics to handle them.

In contrast, #51 seems focused purely on 3., which seems like a much more complicated thing to document and understand :-)

**Comment by aspiers** [comment]
I think I eventually found the answer I was looking for:

`M-x straight-use-package NEW-RECIPE` invokes `(straight--register-recipe recipe)` which puts a new entry for `NEW-RECIPE` into `straight--recipe-cache`.  Once it's in that cache, commands like `straight-rebuild-package` work fine.  So I guess the process for writing a new MELPA recipe (say) is something like:

- Create a new branch in `~/.emacs.d/straight/repos/melpa`
- Put the new recipe in `~/.emacs.d/straight/repos/melpa/recipes/NEW-RECIPE`
- `M-x straight-use-package NEW-RECIPE`
- Test it, and if any issues are found, fix them then jump back to the previous step
- Make a commit in the `melpa` repo and submit a PR

Is that roughly right?

**Comment by aspiers** [comment]
Seems that isn't right, because after this the new entry vanishes from the package completion lists.  Actually I'm totally confused why `straight-get-recipe` filters out any packages in `straight--repo-cache` from the completion list used when called interactively:

    (let* ((package (intern
                     (completing-read
                      "Which recipe? "
                      (cl-remove-if (lambda (pkg)
                                      (gethash pkg straight--repo-cache))
                      (straight-recipes-list sources))

That means it's *only* in the list of completions given by `straight-get-recipe` *before* it's registered.  What's the logic behind that?

**Comment by progfolio** [comment]
> Actually I'm totally confused why straight-get-recipe filters out any packages in straight--repo-cache from the completion list used when called interactively:

See: https://github.com/radian-software/straight.el/issues/976

**Comment by aspiers** [comment]
OK thanks, that's helpful.  The problem with this change to `straight-use-package` is that it assumes that the only use case for that function is installing new packages.  But my use case is different: developing a new recipe and testing it (i.e. reinstalling) until it works.  It seems there is no convenient interactive command for registering a new recipe (or new version thereof) from one of the local recipe repositories such as `~/.emacs.d/straight/repos/melpa/recipes/NEW-RECIPE`.  `straight-register-package` is non-interactive and requires passing the actual recipe, rather than just the package name.  Or am I missing some alternative to `straight-use-package` for this?

IMHO this confusion highlights the need for documenting the recommended process, as despite having spent several hours poring over `straight.el` internals, I still haven't figured out a reliable process.  If I find an issue in the recipe I'm developing, should I remove the installed package somehow and then re-run `straight-use-package`?  Or is there a better way?  Thanks.

**Comment by aspiers** [comment]
Ah, the problem is actually worse than I thought, since it's a change to `straight-get-recipe` not `straight-use-package`, even though the original request in #460 was regarding `straight-use-package`.  I can kind of empathise with the request in #460, even though it causes the problem described above, but it makes zero sense to me for `straight-get-recipe` to stop working once a package is installed.

So I propose moving this filter from `straight-get-recipe` to `straight-use-package`.  (But even then, I'd still be looking for an officially supported way to update `straight--repo-cache` with a new recipe for a given package.)

Thoughts?

**Comment by aspiers** [comment]
I found another issue: it seems that if a recipe is already registered, `(straight-use-package PACKAGE)` will just reuse the existing recipe in the cache.  I think I've finally found a workaround, which is to remove the old recipe from the cache first:

```elisp
(remhash "my-package" straight--recipe-cache)
```

but hopefully there is a more user-friendly way to do this which doesn't rely on knowledge of `straight.el` internals?

**Comment by progfolio** [comment]
Does the version of straight you are using contain the change in https://github.com/radian-software/straight.el/commit/b89a0f700f6d1fe9fe0b0721d7d5f8f50a0520f5 ?

**Comment by aspiers** [comment]
No it didn't, as I was on `master` not `develop`.  Thanks for highlighting that.  I still believe it makes more sense to move the filter to `straight-use-package` however, since I can't think of any good reason why `straight-get-recipe` should stop working for a package once it's installed.  Am I missing something?  If a function with that behaviour is genuinely needed (rather than just as a means for modifying the behaviour in `straight-use-package`), then wouldn't it make more sense to have a separate function called `straight-get-uninstalled-recipe`?

**Comment by progfolio** [comment]
>  I can't think of any good reason why straight-get-recipe should stop working for a package once it's installed

Interactively it works as expected:

```emacs-lisp
(straight-bug-report
  :interactive t
  :post-bootstrap
  ;; install doct
  (straight-use-package 'doct)
  ;; you should still be able to get doct's recipe from the candidate list.
  (call-interactively #'straight-get-recipe))
  ```
  
  > wouldn't it make more sense to have a separate function called straight-get-uninstalled-recipe?
 
It would be more flexible to have `straight-get-recipe` take a generic filter function (rather than the "unflitered" parameter it accepts now). That would cover both cases and future possible cases.

> it seems that if a recipe is already registered, (straight-use-package PACKAGE) will just reuse the existing recipe in the cache.

If I understand correctly, this the main issue.

**Comment by aspiers** [comment]
> Interactively it works as expected

... but not non-interactively by default, unless this new `unfiltered` argument is used.

> It would be more flexible to have `straight-get-recipe` take a generic filter function (rather than the "unflitered" parameter it accepts now). That would cover both cases and future possible cases.

Yes, that sounds great.  But I still think the default behaviour for this function should be unfiltered, because there's nothing in the function name which sets an expectation otherwise.  IIUC, it used to have this unfiltered behaviour, and the only reason it was changed was in order to address a specific use case of `straight-use-package`, even though that's a higher-level function.

> > it seems that if a recipe is already registered, (straight-use-package PACKAGE) will just reuse the existing recipe in the cache.
>
> If I understand correctly, this the main issue.

Not really; it's one particular issue I have when developing new recipes.

**Comment by progfolio** [comment]
>  IIUC, it used to have this unfiltered behaviour, and the only reason it was changed was in order to address a specific use case of straight-use-package, even though that's a higher-level function.

They're both commands. Neither is "higher-level".

> the default behaviour for this function should be unfiltered

I've made the change I described above in #1044. Testing is appreciated.


> > > it seems that if a recipe is already registered, (straight-use-package PACKAGE) will just reuse the existing recipe in the cache.
> > 
> > 
> > If I understand correctly, this the main issue.
> 
> Not really; it's one particular issue I have when developing new recipes.

What I meant was that if interactively evaluating the `straight-use-package` form bypassed the cache and installed the package with the updated recipe, one could edit and evaluate the recipe/form in place.
That should be enough to develop most recipes. Testing for MELPA (or any ELPA) recipe compatibility its own, separate issue.

**Comment by aspiers** [comment]
> They're both commands. Neither is "higher-level".

I meant this in the sense that `straight-use-package` calls `straight-get-recipe`, but the converse is not true.  So the former is "higher" in the call hierarchy.  However I appreciate that despite this they may be considered part of the same abstraction layer.

> I've made the change I described above in https://github.com/radian-software/straight.el/pull/1044. Testing is appreciated.

Works perfectly, thanks!

> What I meant was that if interactively evaluating the straight-use-package form bypassed the cache and installed the package with the updated recipe, one could edit and evaluate the recipe/form in place.  That should be enough to develop most recipes.

It would be sufficient if documented, yes.

> Testing for MELPA (or any ELPA) recipe compatibility is a separate issue.

True.  Ideally both would be documented.  I guess probably just a few lines of text would be sufficient, but I don't currently feel confident enough in my understanding to write it myself.

**Comment by raxod502** [comment]
Added documentation in https://github.com/radian-software/straight.el/pull/1242, let me know what else can be improved.

**Comment by aspiers** [comment]
Thanks a lot @raxod502!
